File: problems/2025/09-04/linked-list-group-reverse/solution.py
from typing import Optional, List


class Node:
    def __init__(self, key: int):
        self.data = key
        self.next: 'Optional[Node]' = None


class Solution:
    def reverseKGroup(self, head: Optional[Node], k: int) -> Optional[Node]:
        if head is None or k <= 1:
            return head

        dummy = Node(0)
        dummy.next = head
        prev_tail: Node = dummy
        curr: Optional[Node] = head

        while curr is not None:
            last_node_of_prev_part: Node = prev_tail
            last_node_of_sublist: Node = curr

            prev: Optional[Node] = None
            i = 0
            # reverse up to k nodes (or remaining nodes if < k)
            while curr is not None and i < k:
                nxt = curr.next
                curr.next = prev
                prev = curr
                curr = nxt
                i += 1

            # connect the previous part with the reversed sublist
            last_node_of_prev_part.next = prev
            # connect the tail of this sublist to the next group's head
            last_node_of_sublist.next = curr
            # move prev_tail to the tail of this reversed group
            prev_tail = last_node_of_sublist

        return dummy.next


# Helpers for quick checks

def build_list(values: List[int]) -> Optional[Node]:
    head: Optional[Node] = None
    tail: Optional[Node] = None
    for v in values:
        n = Node(v)
        if head is None:
            head = tail = n
        else:
            assert tail is not None
            tail.next = n
            tail = n
    return head


def to_array(head: Optional[Node]) -> List[int]:
    out: List[int] = []
    while head is not None:
        out.append(head.data)
        head = head.next
    return out


if __name__ == "__main__":
    s = Solution()

    h = build_list([1, 2, 3, 4, 5, 6])
    h = s.reverseKGroup(h, 2)
    print(to_array(h))  # [2, 1, 4, 3, 6, 5]

    h = build_list([1, 2, 3, 4, 5, 6])
    h = s.reverseKGroup(h, 4)
    print(to_array(h))  # [4, 3, 2, 1, 6, 5]

File: problems/2025/09-04/linked-list-group-reverse/solution.js
function Node(key) {
  this.data = key;
  this.next = null;
}

function reverseKGroup(head, k) {
  if (!head || k <= 1) return head;

  const dummy = new Node(0);
  dummy.next = head;
  let prevTail = dummy;
  let curr = head;

  while (curr) {
    const lastNodeOfPrevPart = prevTail;
    const lastNodeOfSublist = curr;

    let prev = null;
    let i = 0;
    while (curr && i < k) {
      const nxt = curr.next;
      curr.next = prev;
      prev = curr;
      curr = nxt;
      i++;
    }

    lastNodeOfPrevPart.next = prev;
    lastNodeOfSublist.next = curr;
    prevTail = lastNodeOfSublist;
  }

  return dummy.next;
}

// Helpers
function buildList(arr) {
  let head = null, tail = null;
  for (const v of arr) {
    const n = new Node(v);
    if (!head) head = tail = n;
    else { tail.next = n; tail = n; }
  }
  return head;
}

function toArray(head) {
  const out = [];
  while (head) { out.push(head.data); head = head.next; }
  return out;
}

// quick checks
let h = buildList([1,2,3,4,5,6]);
h = reverseKGroup(h, 2);
console.log(toArray(h)); // [2,1,4,3,6,5]

h = buildList([1,2,3,4,5,6]);
h = reverseKGroup(h, 4);
console.log(toArray(h)); // [4,3,2,1,6,5]

